src/System.Globalization.Calendars/tests/CalendarHelpers.cs

```diff
@@ -2,88 +2,209 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
 using Xunit;
 
 namespace System.Globalization.Tests
 {
     public static class CalendarHelpers
     {
-        private static int MaxEra(Calendar calendar) => calendar.GetEra(calendar.MaxSupportedDateTime);
+        public static Calendar[] s_calendars = new Calendar[]
+        {
+            new ThaiBuddhistCalendar(),
+            new JapaneseCalendar()
+        };
+
         private static int MinEra(Calendar calendar) => calendar.GetEra(calendar.MinSupportedDateTime);
+        private static int MaxEra(Calendar calendar) => calendar.GetEra(calendar.MaxSupportedDateTime);
 
-        private static int YearOffset(Calendar calendar) => calendar.GetYear(calendar.MinSupportedDateTime) - 1;
+        private static int MaxRelativeYearInEra(Calendar calendar, int era)
+        {
+            int[] eras = calendar.Eras;
+            if (eras.Length == 1 || era == eras[0])
+            {
+                return MaxDateTimeInEra(calendar, era);
+            }
+            int difference = MaxDateTimeInEra(calendar, era) - MinDateTimeInEra(calendar, era) + 1;
+            if (era != MaxEra(calendar))
+            {
+                difference += 1;
+            }
+            return difference;
+        }
 
-        public static void AddYears_Invalid(Calendar calendar)
+        private static int MinDateTimeInEra(Calendar cal, int era)
         {
-            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddYears(calendar.MaxSupportedDateTime, 1));
-            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddYears(calendar.MinSupportedDateTime, -1));
+            int[] eras = cal.Eras;
+            Assert.InRange(era, 0, eras[0]);
+            if (eras.Length == 1)
+            {
+                return cal.GetYear(cal.MinSupportedDateTime);
```

Comment:
here is some helpers that you may use 

        static int GetMaxGregorianYearInEra(Calendar cal, int era)
        {
            int[] eras = cal.Eras;

            // Assert here ere is less than eras[0]

            if (eras.Length == 1 || era == eras[eras.Length - 1] || era == 0)
            {
                return cal.MinSupportedDateTime.Year;
            }

            return cal.ToDateTime(1, 1, 1, 0, 0, 0, 0, era).Year;
        }
        static int GetMaxCalendarYearInEra(Calendar cal, int era)
        {
            int[] eras = cal.Eras;

            // Assert here ere is less than eras[0]

            if (eras.Length == 1 || era == eras[eras.Length - 1] || era == 0)
            {
                return cal.GetYear(cal.MinSupportedDateTime);
            }

            return cal.GetYear(cal.ToDateTime(1, 1, 1, 0, 0, 0, 0, era));
        }

and look at the places you call any calendar API and try to determine if the API expect calendar date or Gregorian date to use. please let me know if you have any question. and thanks a lot for your effort.